core/src/oracles.ts

```diff
@@ -0,0 +1,135 @@
+import { getContractInterfaceByName } from './contracts';
+import getProvider from './provider';
+import {
+    CollateralConfig,
+    CollateralPriceSourceConfig,
+    OraclePrices,
+    OracleCurrentPriceOnly,
+    OracleCurrentAndNextPrices,
+    CollateralType,
+} from './types';
+import BigNumber from './bignumber';
+import { ethers } from 'ethers';
+import { WAD_NUMBER_OF_DIGITS } from './constants/UNITS';
+import memoizee from 'memoizee';
+import COLLATERALS from './constants/COLLATERALS';
+
+const CACHE_EXPIRY_MS = 24 * 60 * 60 * 1000;
+const getNextOraclePrice = async (
+    oracle: CollateralPriceSourceConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+): Promise<BigNumber> => {
+    /**
+     * Determine if the next price can be fetched from the contract,
+     * If yes, fetch the price by direct memory access (via slot address)
+     * The price is stored in the same slot as a validity marker:
+     *   - split the received value into validity marker part and price part
+     * If the price is valid - return it, otherwise provide a NaN
+     **/
+    if (oracle.type !== 'CurrentAndNextPrice') {
+        return new BigNumber(NaN);
+    }
+    const slot = oracle.nextPriceSlotAddress;
+    const nextPriceFeed = await provider.getStorageAt(oracleAddress, slot);
```

Comment:
If this variable contains both price and the validity, maybe that's how it should be called?
```suggestion
    const nextPriceAndValidityHex = await provider.getStorageAt(oracleAddress, slot);
```